KEP-4724: support node topologies in downward API (#5146)
kubernetes/enhancements: keps/sig-node/4742-node-topology-downward-api/README.md

  within the machine learning framework itself (e.g., Ray, PyTorch Distributed. etc).  Kubernetes provides the
  necessary foundation for these frameworks to operate efficiently, but the frameworks are ultimately responsible
  for leveraging that foundation.
* CNI plugins can leverage topology information to establish more efficient network paths,
  reducing latency and increasing throughput. For instance, a CNI could prioritize connections
  within the same availability zone or rack.
* Distributed stateful applications, such as sharded databases, can use topology information to improve fault tolerance.
  By spreading replicas and traffic across different failure domains (e.g., zones, racks), these applications can achieve
  higher availability and resilience to failures in any one topology.

Today, applications obtain topology information by using an init container to query the Kubernetes API for the underlying node,
extracting the necessary data from the Node resource. As more workloads and use cases benefit from topology awareness,
we aim to simplify access to this information for Pods via the Downward API.

### Goals

> * Values from Node labels `topology.k8s.io/zone`, `topology.k8s.io/region` and `kubernetes.io/hostname` are made

Reviewer: Are you dropping the `*.topology.k8s.io` affordance from this KEP?

Reviewer: On one hand, less scope is more gooder.

On the other hand, I thought it was expressly part of the original KEP that this be more extensible?  Yes, users can install a mutating admission - is that oo heavyweight for this?  How common will extensibility be?

I guess we can wait and see.

Author: Yes, I decided to only include the standard labels for now as it's safer and doing `*.topology.k8s.io/*` doesn't seem that much more extensible than what we doing now. Happy to revisit this after alpha once we get some user feedback.

  available via downward API

> * Additional node labels can be made available via downward API using admission webhooks that mutate `pods/binding`.

Reviewer 2: can you elaborate on this? Is it covered somewhere in the KEP on how to implement this admission webhook?

Author: The implementation of the mutating webhook itself will be very similar to the built-in admission plugin we are going to write, except it's an external mutating webhook (requiring a MutatingWebhookConfiguration) and the copied labels can be customized.

Reviewer 2: The details I cannot find is what this webhook should copy and where. And what will happen automatically after this and how it may conflict with the built-in functionality

### Non-Goals

* Exposing non-standard node labels by default
* Enhnacements to topology-aware scheduling
* Changes to standard topology labels in Kubernetes
* Changes to downward API

## Proposal

Retreival of topology information will be achieved through the existing mechanism of passing down Pod labels using Downward API.

> A new admission plugin, `PodTopologyLabels`, will be added to kube-apiserver. When enabled, it will mutate the `pods/binding` subresource,

Reviewer 2: > When enabled

Just to clarify, it will be enabled whenever the feature gate is turned on, correct? So on all new clusters in a year or so, all Pods will always have their topology labels, correct?

Reviewer 2: in other words, there is nothing that needs configuring. And any Pod, even those that do not use downward API will have their labels populated

Author: That's correct, but I think before Beta, we should have another debate about whether `PodTopologyLabels` is enabled-by-default built-in admission. 

My biggest gripe about this proposal is the idea of cluttering Pod labels by forcing the topology labels on all Pods.

Author: That's correct, but I think before Beta, we should have another debate about whether `PodTopologyLabels` is enabled-by-default built-in admission. 

My biggest gripe about this proposal is the idea of cluttering Pod labels by forcing the topology labels on all Pods. cc @thockin

Reviewer: I hear the gripe - let's push ahead and decide if we need something like a way to "request" such labels be populated (e.g. a field in spec)

Reviewer 2: The idea that all Pods are attributed with the topology info is very attractive for me. So I agree we should continue with this and listen for the feedback.

Reviewer 3: Sounds like there are two separate (yet related) questions:
 - Whether to enable `PodTopologyLabels` the admission by default
 - Whether to allow certain pods to opt in or out this feature.

One potential issue I see is that if we do this by default for all pods, it might cause problems for users/pods that do not want this feature for whatever reason. And we're unlikely to get feedback from them until this is made the default.

Reviewer 4: I think this is somewhat mitigated by the default admission plugin only copying across the three labels described in the 'goals' section (`topology.k8s.io/zone`, `topology.k8s.io/region` and `kubernetes.io/hostname`).

So the only component that could cause this to 'bloat' is the scheduler or any administrator-configured admission plugins, which seems reasonable.

adding topology labels that match those of the target Node. The Binding REST implementation will be updated to copy these labels
from the `pods/binding` subresource to the assigned Pod's labels.

Using the downward API to retrieve topology information will look similar to the following:
```
apiVersion: v1
kind: Pod
metadata:
  generateName: downwardapi-example-
spec:
  containers:
    - name: container
      image: registry.k8s.io/busybox
      command: ["sh", "-c"]
      args:
      - while true; do
          if [[ -e /etc/podinfo/zone ]]; then
            echo -en '\n\n'; cat /etc/podinfo/zone; fi;
          sleep 5;
        done;
      volumeMounts:
        - name: podinfo
          mountPath: /etc/podinfo
  volumes:
    - name: podinfo
      downwardAPI:
        items:
          - path: "zone"
            fieldRef:

>               fieldPath: metadata.labels['topology.kubernetes.io/zone']

Reviewer 2: what is the expected behavior when the label ended up missing?

Reviewer 2: yes, I just want to make sure that the existing behavior on missing labels is OK for this KEP. PRR reviewers may want to know the effect of disabling the feature gate

Author: I believe expected behavior is that it just has an empty value in the Pod, it doesn't prevent Pods from running

Reviewer 2: maybe mention it in PRR when saying what is the outcome for the end user when feature is disabled

Reviewer 5: good suggestion, thanks Sergey

Author: The topology labels would still be available on already running Pods even if the feature is disabled though, updated PRR to make this clearer

* As a database developer, I want to leverage topology information to improve fault tolerance of sharded databases on Kubernetes.
* As a developer of a Kubernetes CNI plugin, I want to pass topology information down to the CNI plugin to optimize data paths for container networks.

### Notes/Constraints/Caveats (Optional)

### Risks and Mitigations

* Scope creep. Allowing additional node information or node label info could
create security issues. This is mitigated by limiting the node labels
to strictly those that are standardized through KEP-1659.

* Exposing sensitive data as node labels to pods. This is mitigated by ensuring
standard topology labels are available to Pods.

* Stale data. Information obtained through node labels is like information
attained through a configmap or secret mounted to a pod, being passed on
creation but not guaranteed to be immutable and thus should be treated as so.

## Design Details

* A built-in Kubernetes admission plugin, `PodTopologyLabels` will be introduced in kube-apiserver
* The `PodTopologyLabels` admission plugin is responsible for mutating `pods/binding` subresource, adding topology labels matching the target Node.

> * `PodTopologyLabels` admission will not override any existing topology labels on Pods.

Reviewer 2: does this one run after all customer's mutating hooks?

Author: It will depend on re-invocation policy https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#reinvocation-policy

* A feature gate, `PodTopologyLabelsAdmission` will be introduced in v1.33. Alpha and disabled by default.
The `PodTopologyLabels` admission plugin can only be set when this feature gate is enabled.

> * The Binding REST implementation will be updated to copy all labels from `pods/binding` subresource into Pods.

Reviewer 2: is update operation to pod/binding force Pod rescheduling?

Author: This isn't adding a new update operation though, it's just adding a mutating to existing operation

> At this point we do not remove labels, it is assumed that any labels in Binding are safe to expose via Downward API.

Reviewer: If any label key in the Binding already exists in the pod, it will be dropped and will NOT overwrite the pod's value.

> At this point we do not remove labels, it is assumed that any labels in Binding are safe to expose via Downward API.

Reviewer 2: one question for the future. If/when we will allow un-scheduling Pods due to things like failed devices, we will need to have a way to override these values. But we will have troubles doing it if automatic labels has no difference from the original ones. Is it too bad to start with overwriting customer labels? If we believe it will break many customers, than we need to come up with the way to differentiate these labels from customers' labels.

Reviewer: This is a really good question, and again exposes a fundamentally missing piece of our APIs.

options:

1) ignore what I said and blast these labels, regardless of what the pod said before.
2) make kubelet set them on the pod when it agrees to admit it (point of no return).
3) store these somewhere else (e.g. status).
4) ?

@andrewsykim @SergeyKanzhelev do you think this question derails the KEP for this release?

Author: We discussed offline, given we are only allowing standard k8s labels, it should be okay to overwrite them during admission

> * For exposing additional node labels, at the discretion of the cluster admin, a mutating admission webhook can be used to mutate labels of the `pods/binding` subresource.

Reviewer 2: so all labels from pods/binding will be available? Just checking

Author: Not just available, but copied to the Pod object. This is under the assumption that nothing is setting labels on pods/binding today and this would be a first use-case for it.

Reviewer 4: @thockin @enj @liggitt are we happy with this behaviour? To summarise:

1) the admission plugin will copy `topology.k8s.io/{zone,region}` + `kubernetes.io/hostname` from the Node object to the Binding during `pods/binding` admission
2) the Binding registry will indiscriminately copy all labels on the `Binding` object to the Pod, regardless of their prefix

This is nice and extensible as it does mean any scheduler (or mutating admission webhook/policy) can mutate `pods/binding` to inject their own arbitrary organisation specific labels (e.g. `example.com/datahall`), but on the other hand could be confusing if we overwrite existing labels on the Pod...

Reviewer 6: https://github.com/kubernetes/enhancements/pull/5146#discussion_r1953511000 feels relevant.

### Test Plan

[X] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

##### Unit tests

Unit tests will be added for:
* New `PodTopologyLabels` admission plugin
* Binding REST implementation

Unit tests will also ensure behavior is exercised when the feature gate is enabled.

##### Integration tests

Integration tests will be added to test the following behavior:
* Pods contain topology labels when `PodTopologyLabels` admission plugin is enabled.
* Topology labels can be expressed in Pod Downward API
* Node labels outside standard topology labels are disallowed

> * Topology labels are empty if underlying Node does not specify a topology

Reviewer 2: can we add a test that ensures that customer can use their own webhook to copy or set those values

Author: Not sure I'd be willing to write an e2e test that involves introducing an external component. If we can get it working with something like MutatingAdmissionPolicy then yes.

Author: Added an e2e test for using MutatingAdmissionPolicy to test custom node labels

Integration tests will also ensure behavior is exercised when the feature gate is enabled.

##### e2e tests

E2E tests will be added to test the following scenarios:
* Pods contain topology labels when `PodTopologyLabels` admission plugin is enabled.
* A Pod using downward API to retrieve the underlying topology information about the Node
* A Pod attempting to use downward API to retrieve Node labels that are not the standard topology labels
* A Pod using downward API on a Node that does not contain any topology information.

E2E tests will also ensure behavior is exercised when the feature gate is enabled.

### Graduation Criteria

#### Alpha

- All standard topology labels can be retrieved using downward API.
- Behavior is implemented behind a feature gate that is off by default.
- Initial unit, integration and e2e tests completed and enabled.

#### Beta

TODO after Alpha.

#### GA

TODO after Beta.

### Upgrade / Downgrade Strategy

Upgrade should be relatively safe because there no behavior changes unless topology labels
are specified in the downward API.

There are some risks to downgrade, for example, if a Pod relies on the topology

> information using downward API, that information could get lost on downgrade.

Reviewer 2: please highlight the fact that already running pods will not be affected. It is about the new Pods utilizing this feature

Author: done

### Version Skew Strategy

There should be no risks with version skew between control plane and nodes since there's no changes made to the kubelet's downward API implementation.
All changes proposed in this KEP are control-plane only.

For n-3 kubelets, old kubelets still support consuming Pod labels via downward API. This capability should be supported as long as the control plane
has this feature enabled.

There may be some risks for version skew with other components (e.g CSI or CNI) if the downward API is used
to pass topology information to those components. Components should ensure they are using a kube-apiserver version
that supports copying topology labels to Pods.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

A feature gate `PodTopologyLabelsAdmission` will be added to the kube-apiserver to enable or disable this feature.

###### How can this feature be enabled / disabled in a live cluster?

- [X] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: `PodTopologyLabelsAdmission`

>   - Components depending on the feature gate:

Reviewer 2: ```suggestion
  - Components depending on the feature gate: `api-server`
```

Author: thanks, done

- [X] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: `PodTopologyLabelsAdmission`
  - Components depending on the feature gate: `kube-apiserver`
- [ ] Other
  - Describe the mechanism:
  - Will enabling / disabling the feature require downtime of the control
    plane?
  - Will enabling / disabling the feature require downtime or reprovisioning
    of a node?

###### Does enabling the feature change any default behavior?

Yes, all Pods will have topology labels copied over.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes it can, but Pods created while the feature is enabled will continue to have topology labels.

###### What happens if we reenable the feature if it was previously rolled back?

New Pods will have topology labels copied over. Older Pods created while the feature was disabled will be missing topology labels.

###### Are there any tests for feature enablement/disablement?

Tests will be added to ensure feature gate works as expected.

### Rollout, Upgrade and Rollback Planning

TODO for Beta.

###### How can a rollout or rollback fail? Can it impact already running workloads?

> Failure on rollout is unlikely since we do not override any existing topolgoy labels on Pods.

Reviewer 5: s/topolgoy/topology/

Author: fixed

Already running workloads that is already copying topology labels to Pods should be untouched by the new admission plugin.

###### What specific metrics should inform a rollback?

> N/A

Reviewer 5: Could a bug in your code cause binding failures? Is there a metric to track binding failures?

Author: added one of the scheduler metrics here

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

Upgrade / rollback will be manually tested before graduation to Beta.

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

No.

### Monitoring Requirements

TODO for GA.

###### How can an operator determine if the feature is in use by workloads?

> Look for topology labels in Pod labels.

Reviewer 5: There is actually no user control over the feature so this question is N/A.

Author: done

###### How can someone using this feature know that it is working for their instance?

- [ ] Events
  - Event Reason:
- [X] API .status
  - Condition name:
  - Other field: check Pod labels
- [ ] Other (treat as last resort)
  - Details:

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

N/A

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

- [ ] Metrics
  - Metric name:
  - [Optional] Aggregation method:
  - Components exposing the metric:

> - [X] Other (treat as last resort)

Reviewer 5: See above - we want to make sure bindings aren't failing, hopefully there is some metric for that (even if it's higher level scheduling metrics)

Author: added a scheduler metric here

  - Details: SLI are not necessary for this admission plugin

###### Are there any missing metrics that would be useful to have to improve observability of this feature?

> N/A

Reviewer 5: see above note

Author: done

No.

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

Negligible increase to Pod size.

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

No.

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

No.

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

No.

### Troubleshooting

###### How does this feature react if the API server and/or etcd is unavailable?

Little impact as this feature is only relevant when scheduling and running Pods, which would be unavailable anyways

###### What are other known failure modes?

Not all Kubernetes clusters have nodes that have topology labels. Requesting topology information
via downward API in these clusters will result in Pods failing to start or empty values returned
via downward API.

###### What steps should be taken if SLOs are not being met to determine the problem?

Revert feature gate and stop consuming downward API.

## Implementation History

> - the Summary, Motivation, Design Details sections being merged, signaling SIG acceptance

Author: done

## Drawbacks

Topology information is a Node-level construct. By implementing this KEP we are allowing information from Nodes
to trickel down to Pods. Expanding this pattern to additional node information in the future can lead to
weaker security boundaries between Pods and Nodes.

## Alternatives

> Allow retreival of arbitrary Node labels from Pods. This has too many securtiy risks as Nodes are often shared by many tenants.

Reviewer 6: ```suggestion

## Expose all node labels via downward API

Allow retrieval of arbitrary Node labels from Pods. This has too many security risks as Nodes are often shared by many tenants.

## Helper controller

Implement a controller that copies selected node labels to Pods - either as the same label exactly, or by
writing a Pod annotation based on that subset of node labels.

The value of the annotation could be a JSON document.

```

Author: I don't think controller was ever viable due to race with scheduler, but I'll add more details

Author: added

Reviewer 6: TY @andrewsykim

transfers and reduce costs. Having an easy way for pods to access these node
topology labels would provide users a straightforward, maintainable way to
optimize their workloads given topology constraints.

While "topology" is usually associated with the physical layout of a cluster,
it can also be used to describe other types of information about the cluster.
This KEP proposes to allow the expansion of the concept of topology to include
user-defined aspects about their cluster nodes, and in turn provide a way for
pods to receive this information.

Workarounds today typically involve using an initContainer to query the
Kubernetes API and then pass data via shared volume to other containers within
the same pod. By comparison, this proposal would reduce the number of service
accounts and API server clients. Another workaround is to use webhooks to inject
labels into pods, but this relies on advance knowledge of where the pod is going
to be scheduled and requires the webhook to be running and available at the time
of pod creation. This proposal would provide an easier way to access node labels
from pods, and would be more efficient than the current workarounds.

### Goals

> * Values from labels `topology.k8s.io/*` and `kubernetes.io/hostname` are made

Reviewer: Is the goal inclusive of storing those labels in the APIserver, such that they are selectable like normal labels?  Or is the goal to ONLY publish via downward API and specifally NOT be selectable?  Or "don't care".  I think we should choose one explicitly.

Author: This was copy over from the old PR, but I propose we don't select against all labels and codify the topologies in downward API, something like this:
```
        - name: ZONE_ENV_VAR
          valueFrom:
            topologyRef:
              path: topology.zone
        - name: REGION_ENV_VAR
          valueFrom:
            topologyRef:
              path: topology.region
```

What do you think?

* Not to update pod labels after the initial node -> pod copy has been made
* Not to make assurances regarding timing and availability of the label beyond
  the initial pod label copy at scheduling time
* Not to make assurances about the immutability of the pod label after the
  initial copy. As with other labels, the pod label can be updated by the user
  after the pod is created.

## Proposal

KEP 1659 defines the following labels: `topology.kubernetes.io/region` and
`topology.kubernetes.io/zone` to be used for topology information. These labels
are useful for pods as well to be able to make application decisions based on
the region or zone the pod is running in. This KEP proposes to make these labels
available to pods while also expanding upon KEP 1659 to allow for user-defined
labels in the `*.topology.kubernetes.io` namespace.

This KEP expands on KEP 1659 in the following ways:

1. Label prefixes of the form `<domain>.topology.kubernetes.io` are allocated
   for use by end users. The Kubernetes project itself will not define any
   labels with this prefix.
2. Labels of the form `<domain>.topology.kubernetes.io/<field>` will be passed
   to pods.

> 3. Labels of the form `topology.kubernetes.io/*` will be passed to pods but will

Reviewer: There was an open question about hostname?  In or out, we should say here.

Author: I think we should include it

Author: Main thing to consider is whether it's redundant with current field ref `spec.nodeName`. I believe there are some cases where `spec.nodeName` may differ from `kubernetes.io/hostname`, but examples are not coming to mind

   continue to be reserved by the Kubernetes project.
4. All labels with `topology.kubernetes.io` and `*.topology.kubernetes.io`
   prefixes should be considered safe for pods and should only contain
   information that pods and containers can safely consume.

The idea is that we will expose those labels from nodes to pods via a literal

> copy from the node. From that point, the topology labels can be used in the same

Reviewer: ...from the node when the pod is scheduled?  Or when the pod is started by kubelet?

If scheduling: Does it become a responsibility of every scheduler to do this, or is it automatic?

If Kubelet: Does it get locked at pod admission-time or at pod (re)start time (i.e. could change if the pod restarts)?

- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

Many workloads benefit from knowing the underlying topology that a Pod is scheduled in.
However, today there is no mechanism in Kubernetes to pass down node topology information to Pods.
The KEP outlines how node topology labels will made available to Pods via the downward API.

Exposing node topology labels through the downward API provides a native method
from which to attain node label values in pods without the need of a sidecar
to extract information.

## Motivation

We’d like to change the runtime behavior of containers based on node labels.
In our case, we’re using a CNI with DaemonSets to perform network setup, and
would like to configure the network differently based on the presence of a node

> label.

Reviewer 6: Who's “our“? Kubernetes project, or one end user?

Reviewer: To be fair - @andrewsykim took over the KEP from https://github.com/kubernetes/enhancements/pull/4747

Reviewer 6: Maybe it's come over as harsh, but I'm keen we don't merge KEPs - which are kind of official documents - where there's ambiguity that could confuse someone. Maybe this is the first KEP they read and it leaves the wrong impression about project-wide opinions.

Reviewer: oh, it's fair comment, just clarifying history

Author: I ported the KEP over and just filled in the PRR parts to avoid missing the PRR freeze :) I will update the other sections soon

This KEP proposes to allow the expansion of the concept of topology to include
user-defined aspects about their cluster nodes, and in turn provide a way for
pods to receive this information.

Workarounds today typically involve using an initContainer to query the
Kubernetes API and then pass data via shared volume to other containers within
the same pod. By comparison, this proposal would reduce the number of service
accounts and API server clients. Another workaround is to use webhooks to inject
labels into pods, but this relies on advance knowledge of where the pod is going
to be scheduled and requires the webhook to be running and available at the time
of pod creation. This proposal would provide an easier way to access node labels
from pods, and would be more efficient than the current workarounds.

### Goals

* Values from labels `topology.k8s.io/*` and `kubernetes.io/hostname` are made
  available via downward API 

### Non-Goals

* Not to expose additional node info outside of topology labels
* Not to pass any additional node labels other than `topology.k8s.io/*` and
  `*.topology.k8s.io/*` to pods

> * Not to update pod labels after the initial node -> pod copy has been made

Reviewer 6: Why's this a non-goal? Seems odd to exclude it; are we saying we never want this?

No.

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

Negligible increase to Pod size.

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

No.

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

No.

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

No.

### Troubleshooting

###### How does this feature react if the API server and/or etcd is unavailable?

Same as any existing field references used with downward API. 

###### What are other known failure modes?

Not all Kubernetes clusters have nodes that have topology labels. Requesting topology information
via downward API in these clusters will result in Pods failing to start or empty values returned
via downward API.

###### What steps should be taken if SLOs are not being met to determine the problem?

Revert feature gate and stop consuming downward API. 

## Implementation History

> - the `Summary`, `Motivation`, `Design Details` sections being merged, signaling SIG acceptance

Reviewer 6: nit: no backticks needed

This isn't implementation history.

## Drawbacks

Topology information is a Node-level construct. By implementing this KEP we are allowing information from Nodes
to trickel down to Pods. Expanding this pattern to additional node information in the future can lead to
weaker security boundaries between Pods and Nodes.

## Alternatives

> 

Reviewer 6: As an alternative (and remember, alternatives are the choices we considered but _didn't_ pick): a controller that copies selected node labels to Pods (either as the same label exactly, or storing an annotation based on that subset of node labels).

This would make the information available but depending on exactly how we describe it, you might need a helper entrypoint to decode it. Also, somebody needs to write the controller and release it.

Allow getting arbitrary Node labels from Pods. This has too many securtiy risks as Nodes are often shared by many tenants.

> 

Reviewer 6: Should add "run a sidecar to fetch this information from the Kubernetes API and make that available to Pods" and "run an init container to fetch this information from the Kubernetes API and make that available to Pods"

Reviewer: My feeling on this KEP is this:  Yes, we want to not do things that have viable workarounds.  We also want to make common things easy, and the more AI users I talk to, the more I hear this come up.  I feel like we've been foot-dragging for a long time on it. 

Im my ledger, this has moved from "uncommon" to "common" over the last 12-18 months.

Reviewer 6: This KEP comes with a lot of non-goals and constraints. **They** concern me; are we painting ourselves into a set of promises that are right for 2025 but won't fit for 2029 and beyond?

Genuinely, if we could staff https://github.com/kubernetes/kubernetes/issues/116993 and boost adoption of a common, ergonomic sidecar: that'd go a long way and it massively cuts the effort we'd need to make to change our opinions about exposing node topology.
Somewhere there must be an AI / ML company with some money or time to invest here. I would _much_ sooner put [my] time into making that time investment happen and go towards the flexible thing, than making the downward API itself more complex.

When I picture updating https://kubernetes.io/docs/concepts/workloads/pods/downward-api/ to explain the new context, I wince. So many sources of information and different constraints we might need to explain; for example, you shouldn't change any topology labels on a running node if we do this.

Region and zone, sure, you ought to power the thing off before you truck it to another datacenter. But other labels might be more dynamic, such as placing sets of nodes into latency groups. A world of indirect coupling where keeping things simple in one place might force complexity in another.

Reviewer: My revised proposal (last comments today) is simpler.  No magic.  No new `*ref` - they are labels the same as any label.

* As a database developer, I want to leverage topology information to improve fault tolerance of sharded databases on Kubernetes.
* As a developer of a Kubernetes CNI plugin, I want to pass topology information down to the CNI plugin to optimize data paths for container networks.

### Notes/Constraints/Caveats (Optional)

### Risks and Mitigations

* Scope creep. Allowing additional node information or node label info could
create security issues. This is mitigated by limiting the node labels
to strictly those that are standardized through KEP-1659.

* Exposing sensitive data as node labels to pods. This is mitigated by ensuring
standard topology labels are available to Pods.

* Stale data. Information obtained through node labels is like information
attained through a configmap or secret mounted to a pod, being passed on
creation but not guaranteed to be immutable and thus should be treated as so.

## Design Details

* A built-in Kubernetes admission plugin, `PodTopologyLabels` will be introduced in kube-apiserver
* The `PodTopologyLabels` admission plugin is responsible for mutating `pods/binding` subresource, adding topology labels matching the target Node.

> * `PodTopologyLabels` admission will overwrite `topology.k8s.io/*` labels on Pods.

Reviewer 2: all of them? Or just those it plans to copy? I think we need to only overwrite what we plan to set - this way customer's mutating webhook is not affected and can copy additional labels

Reviewer 4: The goals section details `topology.k8s.io/zone`, `topology.k8s.io/region` and `kubernetes.io/hostname` are the only label values that will be copied in the *admission plugin* (these will be copied onto the *Binding*).

The *registry* is what is responsible for copying labels from the (mutated) binding onto the Pod resource (at the same time as it persists the assigned `spec.nodeName`).

Based on `The Binding REST implementation will be updated to copy all labels from `pods/binding` subresource into Pods.` (below), the registry will be enhanced to always copy across _all_ labels from the Binding onto the Pod (regardless whether `topology.k8s.io` or `example.com`), however will only *overwrite* the 3 known `zone`, `region` and `hostname` labels (as per `At this point we will overwrite Pod labels in Binding that are allowed to be exposed via Downward API.` below).

I do agree we need to ensure we are water-tight on defining what and where we expose and copy, as service providers will build upon these behaviours quickly.

* A feature gate, `PodTopologyLabelsAdmission` will be introduced in v1.33. Alpha and disabled by default.

> The `PodTopologyLabels` admission plugin can only be set when this feature gate is enabled.

Reviewer 2: ```suggestion
The `PodTopologyLabels` admission plugin will be enabled by default when this feature gate is enabled.
```

No.

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

Negligible increase to Pod size.

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

No.

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

No.

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

No.

### Troubleshooting

###### How does this feature react if the API server and/or etcd is unavailable?

Little impact as this feature is only relevant when scheduling and running Pods, which would be unavailable anyways

###### What are other known failure modes?

Not all Kubernetes clusters have nodes that have topology labels. Requesting topology information
via downward API in these clusters will result in Pods failing to start or empty values returned
via downward API.

###### What steps should be taken if SLOs are not being met to determine the problem?

Revert feature gate and stop consuming downward API.

## Implementation History

> - - `v1.33`: initial KEP is accepeted and alpha implementation is complete

Reviewer 5: nit: formatting here

* For exposing additional node labels, at the discretion of the cluster admin, a mutating admission webhook can be used to mutate labels of the `pods/binding` subresource.

### Test Plan

[X] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

##### Unit tests

Unit tests will be added for:
* New `PodTopologyLabels` admission plugin
* Binding REST implementation

Unit tests will also ensure behavior is exercised when the feature gate is enabled.

##### Integration tests

Integration tests will be added to test the following behavior:
* Pods contain topology labels when `PodTopologyLabels` admission plugin is enabled.
* Topology labels can be expressed in Pod Downward API
* Node labels outside standard topology labels are disallowed
* Topology labels are empty if underlying Node does not specify a topology

Integration tests will also ensure behavior is exercised when the feature gate is enabled.

##### e2e tests

E2E tests will be added to test the following scenarios:

> * Pods contain topology labels when `PodTopologyLabels` admission plugin is enabled.

Reviewer 3: and not contain any additional labels

* A Pod using downward API to retrieve the underlying topology information about the Node

> * A Pod attempting to use downward API to retrieve Node labels that are not the standard topology labels

Reviewer 3: Might be worth adding a test case of consuming topology labels via downward API when the some labels are not set
